client/components/TestPlanVersionsPage/index.jsx

```diff
@@ -65,13 +145,24 @@ const ThemeTableHeader = styled(UnstyledThemeTableHeader)`
     margin: 0 !important;
 `;
 
+// https://stackoverflow.com/a/53215514
+const useForceUpdate = () => {
+    const [, updateState] = React.useState();
+    const forceUpdate = React.useCallback(() => updateState({}), []);
+    return forceUpdate;
+};
+
 const TestPlanVersionsPage = () => {
     const { testPlanDirectory } = useParams();
 
     const { loading, data, error } = useQuery(TEST_PLAN_VERSIONS_PAGE_QUERY, {
         variables: { testPlanDirectory },
         fetchPolicy: 'cache-and-network'
     });
+    const forceUpdate = useForceUpdate();
+
+    const expandedVersionSections = useRef();
+    const toggleVersionSections = useRef();
```

Comment:
I wonder if this could be simplified by having these two values be a single `useState`? You could have `[expandedVersionSectionId, setExpandedVersionSectionId]`. This way you would only need to check if the `expandedVersionSectionId` matches the `testPlanVersion` when looping over for rendering. The `onClick` could run `setExpandedVersionSectionId(testPlanVersion.id)`. The nice thing about using `useState` is that it would automatically trigger a re-render and you wouldn't need the `useForceUpdate` at all.

Reply:
Thanks for the note! Yeah, this is something Paul and I were hashing out together. We started with useState, but found that it was producing stale data, in effect, we would set one of the sections to open, but it wouldn't open, and when logging the values, we would see things set to "false" even right after trying to set them to "true".

The advantage of useRef is that its data can't ever be stale, but the disadvantage is exactly as you noted, you have to introduce a useForceUpdate function to trigger a rerender.

If you have an alternative idea, that would be welcome. We went back and forth a few times trying to figure out the most elegant way to do this.

Reply:
> Thanks for the note! Yeah, this is something Paul and I were hashing out together. We started with useState, but found that it was producing stale data, in effect, we would set one of the sections to open, but it wouldn't open, and when logging the values, we would see things set to "false" even right after trying to set them to "true".

Weird! That sounds like a bug with how the state was being updated. The state will only update after the rerender but it will also trigger a render of the component and its children so any immediate logs will log a stale value but after the render finishes the value will change and a rerender will happen. `useState` theoretically should be perfectly designed for this precise problem.

Anyway, maybe there is something specific about this situation that I'm not seeing with my shallow window in. This could easily be a side effect of the strange array-based interface for `DisclosureComponent`. I would personally like to put this in a tech debt inventory but we don't need to address it now.